File: library-management/csharp/src/LibraryManagement/Book.cs
namespace LibraryManagement;

public class Book
{
    private readonly List<Copy> _copies = new();
    private int _nextCopyId = 1;

    public Book(string title, string author, Isbn isbn)
    {
        Title = title;
        Author = author;
        Isbn = isbn;
    }

    public string Title { get; }
    public string Author { get; }
    public Isbn Isbn { get; }

    public IReadOnlyList<Copy> Copies => _copies;
    public int CopyCount => _copies.Count;

    internal Copy AddCopy()
    {
        var copy = new Copy(_nextCopyId++, Isbn);
        _copies.Add(copy);
        return copy;
    }

    internal bool RemoveOneCopy()
    {
        var copy = _copies.FirstOrDefault(c => c.Status == CopyStatus.Available);
        if (copy is null)
        {
            if (_copies.Count == 0) return false;
            copy = _copies[0];
        }
        _copies.Remove(copy);
        return true;
    }

    internal Copy? FindAvailableCopy() => _copies.FirstOrDefault(c => c.Status == CopyStatus.Available);

    internal Copy? FindReservedCopy() => _copies.FirstOrDefault(c => c.Status == CopyStatus.Reserved);
}

File: library-management/typescript/src/Book.ts
import { Copy } from './Copy.js';
import { CopyStatus } from './CopyStatus.js';
import { Isbn } from './Isbn.js';

export class Book {
  private readonly _copies: Copy[] = [];
  private _nextCopyId = 1;

  constructor(
    public readonly title: string,
    public readonly author: string,
    public readonly isbn: Isbn,
  ) {}

  get copies(): readonly Copy[] { return this._copies; }
  get copyCount(): number { return this._copies.length; }

  addCopy(): Copy {
    const copy = new Copy(this._nextCopyId++, this.isbn);
    this._copies.push(copy);
    return copy;
  }

  removeOneCopy(): void {
    const available = this._copies.findIndex((c) => c.status === CopyStatus.Available);
    const idx = available >= 0 ? available : this._copies.length > 0 ? 0 : -1;
    if (idx >= 0) this._copies.splice(idx, 1);
  }

  findAvailableCopy(): Copy | undefined {
    return this._copies.find((c) => c.status === CopyStatus.Available);
  }

  findReservedCopy(): Copy | undefined {
    return this._copies.find((c) => c.status === CopyStatus.Reserved);
  }
}
